src/zkevm_specs/evm/execution/oog_constant.py

```diff
@@ -0,0 +1,44 @@
+from ...util import FQ
+from ..instruction import Instruction, Transition
+from ..table import CallContextFieldTag
+from ..execution_state import ExecutionState
+from ..opcode import Opcode
+from ...util import N_BYTES_GAS
+
+
+def oog_constant(instruction: Instruction):
+    # retrieve op code associated to oog constant error
+    opcode = instruction.curr.aux_data
+
+    # check gas left is less than const gas required
+    gas_not_enough, _ = instruction.compare(
+        instruction.curr.gas_left, FQ(Opcode(opcode.expr().n).constant_gas_cost()), N_BYTES_GAS
```

Comment:
Here you're taking the opcode constant gas cost from a map defined in the python spec.  In the circuit we'd want to prove that this gas cost is the correct one according to the opcode, so maybe we need to add a fixed table and do a lookup here?

Reply:
Can't we simply copy constrain it or something similar and save a table? I mean since it's fixed should not be an issue. But makes it less complex maybe.

Reply:
i understand copy constrain intend to handle dynamic lookup, not for fixed table . saying different circuit instance may have different copy tables.

Reply:
> Can't we simply copy constrain it or something similar and save a table? I mean since it's fixed should not be an issue. But makes it less complex maybe.

I don't really understand your proposal.  Here we have a single gadget that works for the OOG of several opcodes, so we need a way to prove the correct mapping from opcode to the gas cost via circuit constraints.

Reply:
> > Can't we simply copy constrain it or something similar and save a table? I mean since it's fixed should not be an issue. But makes it less complex maybe.
> 
> I don't really understand your proposal. Here we have a single gadget that works for the OOG of several opcodes, so we need a way to prove the correct mapping from opcode to the gas cost via circuit constraints.

I meant that we can take the cell that stores the gas cost of the opcode and carry it to this gadget via copy-constraint. And so, we save the table.
But depending on when and how the Error will be placed within the trace, we might not be able to do this due to permutation errors.

Reply:
you mean the` copy-constraint` is the plonk  permutation check ,  not the `copy circuit ` we have in the zkevm circuit ? 😄 I mis-understood,  I think the cell location is not fixed, so not easy to apply copy constraint(we can't  predict which step/tx is success or not contains in incoming block ). this is the common reason why evm circuit doesn't use copy constrain IMO.

Reply:
add fixed lookup table in [fddf694](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/266/commits/fddf6945c15240396623234ac5a5aef22ce6606f) and do lookup.